// editorjs-inline-tool/es/tool.js
/**
 * GenericInlineTool returns an EditorJS.InlineTool capable of wrapping a
 * selected text with any given `tagName`.
 *
 * inspired by
 * @see https://github.com/editor-js/marker/blob/c306bcb33c88eaa3c172eaf387fbcd06ae6b297f/src/index.js
 */
const createGenericInlineTool = ({ sanitize, shortcut, tagName, toolboxIcon, }) => {
    var _a;
    return _a = class GenericInlineTool {
            /**
             * @param {{api: object}}  - Editor.js API
             */
            constructor({ api }) {
                this.api = null;
                this.button = null;
                this.tag = null;
                this.iconClasses = null;
                this.api = api;
                /**
                 * Toolbar Button
                 *
                 * @type {HTMLElement|null}
                 */
                this.button = null;
                /**
                 * Tag that should is rendered in the editor
                 *
                 * @type {string}
                 */
                this.tag = tagName;
                /**
                 * CSS classes
                 */
                this.iconClasses = {
                    base: this.api.styles.inlineToolButton,
                    active: this.api.styles.inlineToolButtonActive,
                };
            }
            /**
             * Create button element for Toolbar
             *
             * @return {HTMLButtonElement}
             */
            render() {
                this.button = document.createElement('button');
                this.button.type = 'button';
                this.button.classList.add(this.iconClasses.base);
                this.button.innerHTML = this.toolboxIcon;
                return this.button;
            }
            /**
             * Wrap/Unwrap selected fragment
             *
             * @param {Range} range - selected fragment
             */
            surround(range) {
                if (!range) {
                    return;
                }
                const termWrapper = this.api.selection.findParentTag(this.tag);
                /**
                 * If start or end of selection is in the highlighted block
                 */
                if (termWrapper) {
                    this.unwrap(termWrapper);
                }
                else {
                    this.wrap(range);
                }
            }
            /**
             * Wrap selection with term-tag
             *
             * @param {Range} range - selected fragment
             */
            wrap(range) {
                /**
                 * Create a wrapper for given tagName
                 */
                const strongElement = document.createElement(this.tag);
                /**
                 * SurroundContent throws an error if the Range splits a non-Text node with only one of its boundary points
                 * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/Range/surroundContents}
                 *
                 * // range.surroundContents(span);
                 */
                strongElement.appendChild(range.extractContents());
                range.insertNode(strongElement);
                /**
                 * Expand (add) selection to highlighted block
                 */
                this.api.selection.expandToTag(strongElement);
            }
            /**
             * Unwrap term-tag
             *
             * @param {HTMLElement} termWrapper - term wrapper tag
             */
            unwrap(termWrapper) {
                /**
                 * Expand selection to all term-tag
                 */
                this.api.selection.expandToTag(termWrapper);
                const sel = window.getSelection();
                const range = sel.getRangeAt(0);
                const unwrappedContent = range.extractContents();
                /**
                 * Remove empty term-tag
                 */
                termWrapper.parentNode.removeChild(termWrapper);
                /**
                 * Insert extracted content
                 */
                range.insertNode(unwrappedContent);
                /**
                 * Restore selection
                 */
                sel.removeAllRanges();
                sel.addRange(range);
            }
            /**
             * Check and change Term's state for current selection
             */
            checkState() {
                const termTag = this.api.selection.findParentTag(this.tag);
                this.button.classList.toggle(this.iconClasses.active, !!termTag);
                return !!termTag;
            }
            /**
             * Get Tool icon's SVG
             * @return {string}
             */
            get toolboxIcon() {
                return toolboxIcon;
            }
            /**
             * Sanitizer rule
             * @return {Object.<string>} {Object.<string>}
             */
            static get sanitize() {
                return sanitize;
            }
            /**
             * Set a shortcut
             */
            get shortcut() {
                return shortcut;
            }
        },
        /**
         * Specifies Tool as Inline Toolbar Tool
         *
         * @return {boolean}
         */
        _a.isInline = true,
        _a;
};
export default createGenericInlineTool;
